internal/downloader/download.go

```diff
@@ -20,54 +23,79 @@ import (
 // to hold the latest stable version of kubernetes released
 const KubectlStableURL = "https://storage.googleapis.com/kubernetes-release/release/stable.txt"
 
-// Downloder is an helper class that is used to interact with the
+// Downloder is a helper class that is used to interact with the
 // kubernetes infrastructure holding released binaries and release information
 type Downloder struct {
 }
 
-// UpstreamStableVersion returns the latest version of kubernetes that upstream
-// considers stable
-func (d *Downloder) UpstreamStableVersion() (semver.Version, error) {
-	res, err := http.Get(KubectlStableURL)
+func (d *Downloder) getContentsOfURL(url string) (string, error) {
+	res, err := http.Get(url)
 	if err != nil {
-		return semver.Version{}, err
+		return "", err
 	}
 	if res.StatusCode != http.StatusOK {
-		return semver.Version{},
+		return "",
 			fmt.Errorf(
 				"GET %s returned http status %s",
-				KubectlStableURL,
+				url,
 				res.Status,
 			)
 	}
 
 	v, err := ioutil.ReadAll(res.Body)
 	res.Body.Close()
 	if err != nil {
-		return semver.Version{}, err
+		return "", err
 	}
+	return string(v), nil
+}
 
-	return semver.ParseTolerant(string(v))
+// UpstreamStableVersion returns the latest version of kubernetes that upstream
+// considers stable
+func (d *Downloder) UpstreamStableVersion() (semver.Version, error) {
+	v, err := d.getContentsOfURL(KubectlStableURL)
+	if err != nil {
+		return semver.Version{}, err
+	}
+	return semver.ParseTolerant(v)
 }
 
 // GetKubectlBinary downloads the kubectl binary identified by the given version
 // to the specified destination
 func (d *Downloder) GetKubectlBinary(version semver.Version, destination string) error {
-	downloadURL, err := d.kubectlDownloadURL(version)
-	if err != nil {
-		return err
-	}
+	var firstErr error
 
-	if _, err := os.Stat(filepath.Dir(destination)); err != nil {
-		if os.IsNotExist(err) {
-			err = os.MkdirAll(filepath.Dir(destination), os.ModePerm)
-		}
+	for iter := 0; iter < 3; iter++ {
```

Comment:
nitpicking: make `3` a constant defined somewhere, so that we can easily change it afterwards

Reply:
Done in commit 17fe1e5